presto-main/src/main/java/io/prestosql/sql/analyzer/StatementAnalyzer.java

```diff
@@ -2203,12 +2205,19 @@ private void verifySelectDistinct(QuerySpecification node, List<Expression> outp
                     throw semanticException(EXPRESSION_NOT_IN_DISTINCT, node.getSelect(), "For SELECT DISTINCT, ORDER BY expressions must appear in select list");
                 }
 
-                if (!isDeterministic(expression)) {
+                if (!DeterminismEvaluator.isDeterministic(this::isDeterministic, expression)) {
```

Comment:
This looks strange. I'd change the `DeterminismEvaluator.isDeterministic` method to take a `Function<FunctionCall, FunctionMetadata>` and leave it up to DeterminismEvaluator to call `FunctionMetadata.isDeterministic()`. This callsite would then be:

```java
if (!DeterminismEvaluator.isDeterministic(expression, this::getFunctionMetadata)) {
   ...
}
```

And the function below:

```java
private FunctionMetadata getFunctionMetadata(FunctionCall functionCall)
{
    ResolvedFunction resolvedFunction = analysis.getResolvedFunction(functionCall);
    verify(resolvedFunction != null, "function has not been analyzed yet: %s", functionCall);
    return metadata.getFunctionMetadata(resolvedFunction);
}
```